scalesetoperator/scalesetoperator.go

```diff
@@ -0,0 +1,153 @@
+package scalesetoperator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/actions-runner-controller/actions-runner-controller/github"
+	"github.com/actions-runner-controller/actions-runner-controller/pkg/github/runnermanager"
+	"github.com/actions-runner-controller/actions-runner-controller/scalesetlistener"
+	"github.com/go-logr/logr"
+)
+
+type Operator struct {
+	l          *scalesetlistener.Listener
+	logger     logr.Logger
+	maxRunners int
+
+	mu      sync.Mutex
+	started bool
+}
+
+func New(l *scalesetlistener.Listener, logger logr.Logger, maxRunners int) *Operator {
+	return &Operator{
+		l:          l,
+		logger:     logger,
+		maxRunners: maxRunners,
+	}
+}
+
+func (op *Operator) RunJobOperator(ctx context.Context) error {
+	op.mu.Lock()
+	if op.started {
+		op.mu.Unlock()
+		return errors.New("operator started")
+	}
+	op.started = true
+	op.mu.Unlock()
+
+	listenerErr := make(chan error)
+	go func() {
+		listenerErr <- op.l.Run(ctx)
+	}()
+
+	jobOperator := jobOperator{
+		actionsClient:  op.l.ActionsServiceClient(),
+		runnerScaleSet: op.l.RunnerScaleSet(),
+		logger:         op.logger,
+		buffer:         make(chan struct{}, op.maxRunners),
+	}
+
+	runnerJobList, err := runnermanager.GetScaleSetJobs(ctx, jobOperator.runnerScaleSet, "default")
+	if err != nil {
+		return fmt.Errorf("failed to fetch the state: %v", err)
+	}
+	prevState := len(runnerJobList.Items)
+	go jobOperator.consume(ctx)
+
+	stream := op.l.MessageStream()
+	for {
+		select {
+		case err := <-listenerErr:
+			return err
+		case <-ctx.Done():
+			close(jobOperator.buffer)
```

Comment:
I believe we will want to have this statement in the err case above as well. Otherwise we could end up in a deadlock with the `consume` goroutine.

Reply:
Good catch!